gui/wxpython/gui_core/menu.py

```diff
@@ -305,3 +307,68 @@ def OnItemSelected(self, node):
             label = data['description']
 
         self.showNotification.emit(message=label)
+
+
+class RecentFilesMenu:
+    """Add recent files history menu"""
+
+    recent_files = '.recent_files'
+
+    def __init__(self, app_name, parent_menu, pos, history_len=10):
+        self._history_len = history_len
+        self._parent_menu = parent_menu
+        self._pos = pos
+
+        self.file_requested = Signal('FileRequested')
+
+        self.filehistory = wx.FileHistory(maxFiles=history_len)
+        # Recent files path stored in GRASS GIS config dir in the
+        # .recent_files file in the group by application name
+        self.config = wx.FileConfig(
+            style=wx.CONFIG_USE_LOCAL_FILE,
+            localFilename=os.path.join(
+                utils.GetSettingsPath(), self.recent_files,
+            ),
+        )
+        self.config.SetPath(strPath=app_name)
+        self.filehistory.Load(self.config)
+
+        self.recent = wx.Menu()
+        self.filehistory.UseMenu(self.recent)
+        self.filehistory.AddFilesToMenu()
+
+        # Show recent files menu if count of items in menu > 0
+        if self.filehistory.GetCount() > 0:
+            self._InsertMenu()
+
+    def _InsertMenu(self):
+        self._parent_menu.Insert(
+            pos=self._pos, id=wx.ID_ANY, text=_('&Recent Files'),
+            submenu=self.recent,
+        )
+        self.recent.Bind(
+            wx.EVT_MENU_RANGE, self._OnFileHistory,
+            id=wx.ID_FILE1, id2=wx.ID_FILE + self._history_len,
+        )
+
+    def _ShowMenu(self):
+        if self.filehistory.GetCount() == 0:
+            self._InsertMenu()
+
+    def _OnFileHistory(self, event):
+        file_exists = True
+        file_num = event.GetId() - wx.ID_FILE1
+        path = self.filehistory.GetHistoryFile(file_num)
+        if not os.path.exists(path):
+            self.filehistory.RemoveFileFromHistory(i=file_num)
+            file_exists = False
```

Comment:
I guess in ideal case, files which don't exist would not be in the menu in the first place. Or? Hard to say what is more expected in this case. This (the current implementation) seems like an advantageous implementation because it checks only when really needed and we need to check it later on anyway.

Reply:
All right, I fixed it.